### PR TITLE
parsing: Fix dropped attributes after a '-' or '+'

### PR DESCRIPTION
**Pull request by Julow**
The special prefix operators [-], [-.], [+] and [+.] could rewrite:

    -(1 [@foo])

into the constant:

    -1

which drops the attribute. This adds the attribute check, the above
example is now parsed to the correct AST:

    ~-(1 [@foo])

An alternative implementation would be to drop the rewrite of constants from the parser and instead to perform it during a later phase.

Spotted by @gasche in https://github.com/ocaml/ocaml/pull/12968#discussion_r1513425531

### CODE REVIEW

File: testsuite/tests/parsing/prefix_op.ml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+(* TEST_BELOW
+(* Blank lines added here to preserve locations. *)
+
+
+
+
+*)
```
Comment (gasche) [inline testsuite/tests/parsing/prefix_op.ml]: This is a new test, so you don't need those blank lines -- I guess you copied it from somewhere else.

File: testsuite/tests/parsing/prefix_op.ml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
++(1[@foo]);;
+-(1.0[@foo]);;
++(1.0[@foo]);;
+-.(1.0[@foo]);;
++.(1.0[@foo]);;
+
+~+(2[@foo]);;
+
+(* TEST
+ flags = "-dparsetree";
```
Comment (gasche) [inline testsuite/tests/parsing/prefix_op.ml]: I would also use `-dno-locations` to make the output more pleasant to read.

File: testsuite/tests/parsing/prefix_op.ml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ ocamlc_byte_exit_status = "2";
```
Comment (gasche) [inline testsuite/tests/parsing/prefix_op.ml]: Instead of having the compiler fail, you can use `-stop-after parsing`.

**Review by gasche — COMMENTED** [review]
This looks reasonable to me, thanks for the fix! All my comments ended up being about the test script, which is more complex than it could be.

File: testsuite/tests/parsing/prefix_op.ml (hunk 1 above)
Comment (Julow) [reply testsuite/tests/parsing/prefix_op.ml]: Fixed! Indeed, I didn't know how to write a test and so took an other one as a base. Looks much better with `-dno-locations -stop-after parsing`!

**Review by gasche — APPROVED** [review]

**Comment by Julow** [comment]
I investigated why we need this special case instead of always applying the `~+` and `~-` operators. The reason is int32 and int64: `-1l` is an int32 while `~-1l` is a typing error.

**Pull request merged**